src/modules/commander/Commander.cpp

```diff
@@ -4123,36 +4124,92 @@ Commander::offboard_control_update()
 
 void Commander::esc_status_check(const esc_status_s &esc_status)
 {
-	char esc_fail_msg[50];
-	esc_fail_msg[0] = '\0';
+	if ((esc_status.esc_count > 0) && (hrt_elapsed_time(&esc_status.timestamp) < 700_ms)) {
+		// Some DShot ESCs are unresponsive for ~550ms during their initialization, so we use a timeout higher than that
 
-	int online_bitmask = (1 << esc_status.esc_count) - 1;
+		char esc_fail_msg[50];
+		esc_fail_msg[0] = '\0';
 
-	// Check if ALL the ESCs are online
-	if (online_bitmask == esc_status.esc_online_flags) {
-		_status_flags.condition_escs_error = false;
-		_last_esc_online_flags = esc_status.esc_online_flags;
+		int online_bitmask = (1 << esc_status.esc_count) - 1;
 
-	} else if (_last_esc_online_flags == esc_status.esc_online_flags || esc_status.esc_count == 0)  {
+		// Check if ALL the ESCs are online
+		if (online_bitmask == esc_status.esc_online_flags) {
 
-		// Avoid checking the status if the flags are the same or if the mixer has not yet been loaded in the ESC driver
+			_status_flags.condition_escs_error = false;
+			_last_esc_online_flags = esc_status.esc_online_flags;
 
-		_status_flags.condition_escs_error = true;
+		} else if (_last_esc_online_flags == esc_status.esc_online_flags)  {
+
+			// Avoid checking the status if the flags are the same or if the mixer has not yet been loaded in the ESC driver
+
+			_status_flags.condition_escs_error = true;
+
+		} else if (esc_status.esc_online_flags < _last_esc_online_flags) {
+
+			// Only warn the user when an ESC goes from ONLINE to OFFLINE. This is done to prevent showing Offline ESCs warnings at boot
+
+			for (int index = 0; index < esc_status.esc_count; index++) {
+				if ((esc_status.esc_online_flags & (1 << index)) == 0) {
+					snprintf(esc_fail_msg + strlen(esc_fail_msg), sizeof(esc_fail_msg) - strlen(esc_fail_msg), "ESC%d ", index + 1);
+					esc_fail_msg[sizeof(esc_fail_msg) - 1] = '\0';
+				}
+			}
 
-	} else if (esc_status.esc_online_flags < _last_esc_online_flags) {
+			mavlink_log_critical(&_mavlink_log_pub, "%soffline", esc_fail_msg);
 
-		// Only warn the user when an ESC goes from ONLINE to OFFLINE. This is done to prevent showing Offline ESCs warnings at boot
+			_last_esc_online_flags = esc_status.esc_online_flags;
+			_status_flags.condition_escs_error = true;
+		}
+
+		_status_flags.condition_escs_failure = false;
 
 		for (int index = 0; index < esc_status.esc_count; index++) {
-			if ((esc_status.esc_online_flags & (1 << index)) == 0) {
-				snprintf(esc_fail_msg + strlen(esc_fail_msg), sizeof(esc_fail_msg) - strlen(esc_fail_msg), "ESC%d ", index + 1);
-				esc_fail_msg[sizeof(esc_fail_msg) - 1] = '\0';
+
+			if (esc_status.esc[index].failures > esc_report_s::FAILURE_NONE) {
+				_status_flags.condition_escs_failure = true;
+
+				if (esc_status.esc[index].failures != _last_esc_failure[index]) {
+
+					if (esc_status.esc[index].failures & esc_report_s::FAILURE_OVER_CURRENT_MASK) {
```

Comment:
These aren't really generic failures (not used by UAVCAN, DSHOT, etc), perhaps the error message should be confined to the specific module?

Reply:
Mh not really, I've seen this as a recurring pattern where people with proprietary ESCs ended up doing the logic in the driver (and the failures ended up to be almost the same). I think that this is a good first step towards supporting such edge cases

Reply:
Alright, let's see how things continue to evolve.
I'm trying to fine the line where commander is tied in such that it's aware of the components and high level health status, but doesn't have to have specific intimate knowledge unless it's actually going to do something with that information.